lib/reek/smells/instance_variable_assumption.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+require_relative 'smell_detector'
+require_relative 'smell_warning'
+
+module Reek
+  module Smells
+    #
+    # The +InstanceVariableAssumption+ class is responsible for
+    # detecting directly access of instance variables in a class
+    # that does not define them in its initialize method.
+    #
+    class InstanceVariableAssumption < SmellDetector
+      def self.contexts
+        [:class]
+      end
+
+      # Checks +klass+ for instance
+      # variables assumptions.
+      #
+      # @return [Array<SmellWarning>]
+      #
+      def sniff(ctx)
+        method_expressions = ctx.node_instance_methods
+        assumptions = InstanceVariableAssumptionsExtractor.new(method_expressions).extract
+
+        return [] if assumptions.empty?
+
+        build_smell_warning(ctx, assumptions)
+      end
+
+      private
+
+      def build_smell_warning(ctx, assumptions)
+        message = "assumes too much for instance variable #{assumptions.join(', ')}"
+
+        [smell_warning(
+          context: ctx,
+          lines: [ctx.exp.line],
+          message: message,
+          parameters: { assumptions: assumptions })]
+      end
+
+      #
+      # The InstanceVariableAssumptionsExtractor is the class responsible for
+      # extracting the assumptions from a method expression
+      #
+      class InstanceVariableAssumptionsExtractor
```

Comment:
This feels a little overengineered here. I mean, the way it is right now the smell detector itself is not much but an empty shell not doing much. Do we really need this class? How about dissolving it into the smell detector?